hdfs-executor/src/main/java/org/apache/mesos/hdfs/executor/NameNodeExecutor.java

```diff
@@ -50,37 +73,207 @@ public static void main(String[] args) {
+    Runnable r = new Runnable() {
+      public void run() {
+        try {
+          initNameNode(driver, nameNodeTask.getTaskInfo().getName() + ".hdfs.mesos");
+        } catch (Exception ex) {
+          // Failure to start a NameNode on a NameNodeExecutor is catastrophic.
+          FailureUtils.exit("Failed to launch Namenode", HDFSConstants.NAMENODE_EXIT_CODE);
+        }
+      }
+    };
+
+    new Thread(r).start();
+
+    // Lie to Mesos.  Tell it the NameNode Task is running, but we track it's actual
+    // "uninitialized" status in the labels.  The Scheduler depends on these labels
+    // for determing when it should move to the next phase of its state machine.
+    TaskStatus status = StatusFactory.createNameNodeStatus(
+        nameNodeTask.getTaskInfo().getTaskId(),
+        false);
+
+    log.info("Sending status update: " + status);
+    driver.sendStatusUpdate(status);
+  }
+
+  private void launchZKFCTask(final ExecutorDriver driver, final TaskInfo taskInfo) {
+    Task task = new Task(taskInfo);
+    log.info("Launching ZKFC Task: " + task);
+    zkfcNodeTask = task;
+    initZKFCNode(driver);
+  }
+
+  private void initNameNode(ExecutorDriver driver, String dnsName) throws Exception {
+    ensureStatusPathExists();
+    waitDnsResolution(dnsName);
+
+    InterProcessMutex lock = new InterProcessMutex(curatorClient, getStatusPath());
+    if (lock.acquire(HDFSConstants.ZK_MUTEX_ACQUIRE_TIMEOUT_SEC, TimeUnit.SECONDS)) {
+      try {
+
+        // In order to start a set of NameNodes, one must first be formatted.  Other
+        // NameNodes then bootstrap off that node or others which have already bootstrapped.
+        // All Namenoedes are started simultaneously.  Their startups are intentionally
+        // serialized through the mutex acquired above.  The value stored in the Znode which
+        // is used as a mutex indicates whether or not any NameNode has ever been formatted.
+        // The first NameNode to acquire the mutex and find that no NameNode has ever been
+        // formatted, formats itself.  All others bootstrap.
+        if (!nameNodeQuorumInitialized()) {
+          formatNameNode(driver);
+          curatorClient.setData()
+            .forPath(
+                getStatusPath(),
+                HDFSConstants.NN_STATUS_INIT_VAL.getBytes(Charset.forName("UTF-8")));
+        } else {
+          bootstrapNameNode(driver);
+        }
+      } finally {
+        lock.release();
+      }
+    } else {
+      throw new Exception("Failed to initialize NameNode status.");
+    }
+  }
+
+  private void initZKFCNode(ExecutorDriver driver) {
+    if (!processRunning(zkfcNodeTask)) {
+      startProcess(driver, zkfcNodeTask);
+    }
+
+    TaskStatus status = StatusFactory.createRunningStatus(
+        zkfcNodeTask.getTaskInfo().getTaskId());
+    driver.sendStatusUpdate(status);
+  }
+
+  private void ensureStatusPathExists() {
+    try {
+      EnsurePath ensurePath = new EnsurePath(getStatusPath());
```

Comment:
EnsurePath is deprecated.  notes say: Prefer CuratorFramework.create().creatingParentContainersIfNeeded() or CuratorFramework.exists().creatingParentContainersIfNeeded()